ユーザー管理機能の実装

tubuyakisyun opened the pull request:
# what
***ユーザー管理機能***の実装の**3回目修正**を行いました

# why
最終課題　**furima** アプリの実装を行うため（3回目の修正です）



## プルリクエストへ記載するgyazo(**3回目の修正後**)

-  ### [必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画](https://gyazo.com/49ad2ef7b42139c0cb6426159f766e27)

- ### [入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画](https://gyazo.com/518032568fd4e42be31bd84ad13e2b78)
 
- ### [必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画](https://gyazo.com/c337bf94c0353df3fb232676e111a5d7)

- ### [入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画](https://gyazo.com/0a8dcba6142bbee58e5ecf01b831299b)

- ### [ログアウトに成功する動画](https://gyazo.com/daea9e7ff4bbb607a68362f9a29f72e7)
 
- ### [テストに成功する画像](https://gyazo.com/d6182de997c5083fbd02d13c6e742d45)

tubuyakisyun committed:
deviseのインストール

tubuyakisyun committed:
usersテーブルの作成

tubuyakisyun committed:
user バリデーションの記述を行う

tubuyakisyun committed:
要素ファイルの移動を行う

tubuyakisyun committed:
ユーザー登録ファイルのnew.html.erbにnameの変換を行い、登録画面表示

tubuyakisyun committed:
ログイン画面に遷移するように実装し,ユーザーの名前をDBへ保存できるように設定

ログイン画面に遷移するように実装し,ユーザーの名前をDBへ保存できるように設定

tubuyakisyun committed:
不要なER図の削除を行う

tubuyakisyun committed:
Gemfile内で、Gemであるpry-railsを追加

tubuyakisyun committed:
items コントローラを作成する

tubuyakisyun committed:
ログインの有無で表示が変わるように分岐

tubuyakisyun committed:
ユーザーログイン機能

tubuyakisyun committed:
RSpecの準備を行う

必要なGemをインストール
テストの実行ログのフォーマットを整えるように設定
FactoryBotに必要な値を指定

tubuyakisyun committed:
gemファイルにテスト用のライブラリーを記入する

tubuyakisyun committed:
FactoryBotの準備を行う

使用するFactoryBotのファイルを生成
ユーザー情報をFactoryBotで指定
テストを記述するファイルを用意
テストコードのサンプルとして用意されているものを削除

tubuyakisyun committed:
Userモデルをテスト

値が空ではDBに保存できないこと
passwordカラムは、文字数や確認用との不一致によってはDBに保存できないこと
すでにDBに保存されているemailは保存できないこと

tubuyakisyun committed:
Rubocopを実行

ushi10 commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,22 @@
 class ApplicationController < ActionController::Base
+  before_action :authenticate_user!, except: [:show, :index]
```
```ruby
before_action :authenticate_user!, except: [:show, :index]
```
こちらですが、ItemsControllerに記述を移動させておきましょう。
その際は、showアクションを取り外しておきましょう。

ApplicationController内に記載をしてしまうと、複数のコントローラーを作成した際、そちらのindexアクションが動いた時にも読み込まれてしまうためです。
showアクションを削除しておく視点としては、現時点ではまだshowアクションを実装していないためです。
実装タイミングに入りましたら記載をお願いいたします。

tubuyakisyun replied:
- before_action :authenticate_user!, except: [:show, :index] を　ItemsControllerに移動しました
- showアクションを取り外しました

ushi10 commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,22 @@
+      :sign_up,
+      keys: [
+        :nickname,
+        :email,
+        :password,
```
```Ruby
:email,
:password,
```
こちらの記述は削除しておきましょう。
emailとpasswordはdeviseがデフォルトで設定をしてくれているため不要です。

tubuyakisyun replied:
- emailとpasswordの記述を削除しました

ushi10 commented on app/models/user.rb:
```diff
@@ -0,0 +1,16 @@
+  validates :first_name, presence: true
+  validates :last_name_kana, presence: true
+  validates :first_name_kana, presence: true
+
```
以下の実装条件を満たすようなバリデーション、正規表現が不足しております。
そのため、実装をお願いいたします。
```
- パスワードは、半角英数字混合での入力が必須であること
- お名前(全角)は、全角（漢字・ひらがな・カタカナ）での入力が必須であること。
- お名前カナ(全角)は、全角（カタカナ）での入力が必須であること。
```
正規表現を活用すると実装がスムーズにできますので、応用カリキュラムの正規表現カリキュラムを見返したり、外部検索を駆使したりとしながら考えてみましょう！

tubuyakisyun replied:
- 実装条件を満たすようなバリデーションを実装しました

ushi10 commented on app/models/user.rb:
```diff
@@ -0,0 +1,16 @@
+  validates :last_name_kana, presence: true
+  validates :first_name_kana, presence: true
+
+  has_many :items,          foreign_key: :user_id, dependent: :destroy
+  has_many :purchases,      foreign_key: :user_id, dependent: :destroy
```
```Ruby
  has_many :items,          foreign_key: :user_id, dependent: :destroy
  has_many :purchases,      foreign_key: :user_id, dependent: :destroy
```
こちらのアソシエーションは現段階ではコメントアウト、ないしは削除をしておきましょう。
現段階で差分が出てしまうと、商品出品機能等の際に差分が上がらず正確なレビューができなくなってしまうためです。

tubuyakisyun replied:
- コメントアウトしました

ushi10 commented on app/views/layouts/application.html.erb:
```diff
@@ -1,15 +1,17 @@
 <!DOCTYPE html>
 <html>
-  <head>
```
こちらのファイルにて削除をしていただいた部分はどれも必要な記述ですので、元に戻していただきますようお願いいたします。

tubuyakisyun replied:
##- 削除したファイルを元に戻し、一応修正し、コメントアウトしておきます

ushi10 commented on app/views/shared/_header.html.erb:
```diff
@@ -0,0 +1,29 @@
+      <li><%= link_to 'ブランド', "#", class: "brand-list" %></li>
+    </ul>
+    <ul class='lists-right'>
+      <%# deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```
18行目は課題説明文のため削除をしておきましょう。
このようなものが残っている状態でレビュー提出をしたり企業様に提出したりすることは相応しく無いためです。

ushi10 commented on app/views/shared/_header.html.erb:
```diff
@@ -0,0 +1,29 @@
+      <li><%= link_to 'ログイン', new_user_session_path, class: "login" %></li>
+      <li><%= link_to '新規登録', new_user_registration_path, class: "sign-up" %></li>
+      <% end %>
+      <%# //deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```
26行目の記述も先述と同様の理由により削除をしておきましょう。

ushi10 commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,17 @@
+    end
+    nickname { Faker::Name.name }
+    email { Faker::Internet.free_email }
+    password = Faker::Internet.password(min_length: 6)
```
```Ruby
password = Faker::Internet.password(min_length: 6)
```
こちらの記述ですと、稀に英字のみ、もしくは数字のみのパスワードが生成されてしまいます。
このような値は今回のfurimaには適切では無いため、常に英数字混合のパスワードが生成されるよう記述を変更してみましょう。

ushi10 commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,90 @@
+        @user.valid?
+        expect(@user.errors.full_messages).to include('Email is invalid')
+      end
+    end
```
モデルファイル内のuser.rbにバリデーション、正規表現を加えることによりテスト内容が増えていきます。
実装条件や追加した内容に則してテストコードを追記していきましょう。

ushi10 reviewed (commented):
レビュー依頼ありがとうございます！コード、挙動確認動画ともに拝見させていただきました。
9点コメントいたしましたので、ご確認をお願いいたします。
修正頂いた後、コミットプッシュを確認いただき、再度レビュー依頼のご提出をお願い致します！
また、修正後最新のテストコード結果のキャプチャ画像の提出もお願いいたします。

tubuyakisyun committed:
ItemsControllerに記述を移動

tubuyakisyun committed:
emailとpasswordはdeviseがデフォルトで設定をしてくれているため、記述を削除する

tubuyakisyun committed:
お名前(全角)は、全角（漢字・ひらがな・カタカナ）での入力が必須であること。 お名前カナ(全角)は、全角（カタカナ）での入力が必須であること。

tubuyakisyun committed:
バリデーション設定

tubuyakisyun committed:
実装条件を満たすようなバリデーションを実装しました

tubuyakisyun committed:
password,first_name,last_name,first_name_kana,last_name_kanaのテストコードを追加しました

パスワードは、半角英数字混合での入力が必須であること
お名前(全角)は、全角（漢字・ひらがな・カタカナ）での入力が必須であること。
お名前カナ(全角)は、全角（カタカナ）での入力が必須であること。

tubuyakisyun committed:
[]を削除

tubuyakisyun committed:
現段階ではコメントアウト

こちらのアソシエーションは現段階ではコメントアウトしました
現段階で差分が出てしまうと、商品出品機能等の際に差分が上がらず正確なレビューができなくなってしまうためです。

tubuyakisyun committed:
コメントアウト部分はもしかして、削除していいでしょうか

tubuyakisyun committed:
コメントアウトを削除

tubuyakisyun committed:
常に英数字混合のパスワードが生成されるよう記述

tubuyakisyun committed:
user モデルのバリデーション似合うテキストコードに修正する

tubuyakisyun committed:
常に英数字混合のパスワードが生成されるよう記述

こちらはコンソールでFactoryBotが使用できることを確認した

tubuyakisyun committed:
コンフリクトを解決

tubuyakisyun committed:
Update users.rb

tubuyakisyun committed:
RailsのDEPRECATION WARNINGを消す

tubuyakisyun committed:
email validetesを設定

yuri240 commented on app/models/user.rb:
```diff
@@ -0,0 +1,28 @@
+  with_options presence: true do
+      validates :nickname
+      validates :birth_date
+    with_options format: { with:/\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i }, length: { maximum: 255 }, uniqueness: true  do
+      validates :email 
+    end
```
```ruby
    with_options format: { with:/\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i }, length: { maximum: 255 }, uniqueness: true  do
      validates :email 
    end
```

emailのバリデーションは、
deviseのデフォルトで設定されているものを使用しているため
9~11行目の記述は削除していただけると良いか存じます。

tubuyakisyun replied:
#### **削除いたしました。**

yuri240 commented on app/models/user.rb:
```diff
@@ -0,0 +1,28 @@
+      validates :email 
+    end
+    with_options format: {with: /(?=.*[a-zA-Z])(?=.*\d)[a-zA-Z\d]{6,}/, message: "is invalid. Input half-width alphanumeric."} do
+      validates :encrypted_password
+      validates :password
+      validates :password_confirmation
```
パスワードのバリデーションについて、
`encrypted_password`と`password_confirmation`は不要なため
削除しておきましょう！

tubuyakisyun replied:
### **削除いたしました**

yuri240 commented on app/models/user.rb:
```diff
@@ -0,0 +1,28 @@
+      validates :password
+      validates :password_confirmation
+    end
+    with_options format: {with: /\A[ぁ-んァ-ン一-龥]/, message: "is invalid. Input full-width characters."} do
```
```
/\A[ぁ-んァ-ン一-龥]/
```
こちら「佐々木」などである「々」が入力された際に、
弾かれないよう正規表現を下記のように修正しましょう。
↓
```
/\A[ぁ-んァ-ヶ一-龥々ー]+\z/
```

【補足】
・長音記号(ー)も考慮（龥々ー）
・文字列の最後まで考慮（+\z）

tubuyakisyun replied:
#### 有難うございます！　修正致しました！

yuri240 commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,166 @@
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Password is invalid. Input half-width alphanumeric.", "Password confirmation is invalid. Input half-width alphanumeric.")
+      end
+      it 'お名前（全角）は全角（漢字・ひらがな・カタカナ）での入力しないと登録できない' do
+        @user.first_name = 'nana'
+        @user.last_name = '鈴木'
+        @user.valid?
+        expect(@user.errors.full_messages).to include("First name is invalid. Input full-width characters.")
+      end
```
128行目のテストについてですが、
以下の２つに分けてテストを実行しましょう！

```
・first_nameは全角（漢字・ひらがな・カタカナ）での入力しないと登録できない
・last_nameは全角（漢字・ひらがな・カタカナ）での入力しないと登録できない
```


理由：
1つのテストに対して1つの期待するエラーメッセージを設定する原則があるため。

また、これ以降の134~163行目についても同様に
修正をお願いいたします。



>#### **ありがとうございます、修正いたいしました**

tubuyakisyun replied:
>#### **ありがとうございます、修正いたいしました**

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。

追加で４点コメントさせていただきましたので、ご確認お願いします。

tubuyakisyun committed:
emailのバリデーションは、 deviseのデフォルトで設定されているものを使用しているため 9~11行目の記述は削除していただけると良いか存じます。

tubuyakisyun committed:
パスワードのバリデーションについて、 encrypted_passwordとpassword_confirmationは不要なため

tubuyakisyun committed:
「佐々木」などである「々」が入力された際に、 弾かれないよう正規表現を下記のように修正

tubuyakisyun committed:
・first_nameは全角（漢字・ひらがな・カタカナ）での入力しないと登録できない ・last_nameは全角（漢字・ひらがな・カタカナ）での入力しないと登録できないテストコードを実行する

junyakaneko67 commented:
レビュー依頼ありがとうございます！

ただ申し訳ございません、最新修正分がまだ反映されていないようです。
恐れ入りますが、commitやpush忘れなどをご確認いただき、再度プルリクエストをご提出いただけないでしょうか？

宜しくお願い致します

- #### 再度プルリクエストを提出致しました！！！ よろしくお願いします。

tubuyakisyun committed:
Update user_spec.rb

yutamasumoto reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます！実装お疲れ様でした！
このブランチをマージし、masterブランチでプルを行いましょう！
機能が反映されていることを確認し、引き続き実装頑張っていきましょう！


### どうもありがとうございました！
